Compact the pdata to make active droplets contiguous

sjsprecious opened the pull request:
This PR compacts the **pdata** to make sure that the active droplets for each MPI rank is contiguous after calling the **droplet_driver** subroutine. Therefore, there will be no *holes* in the **pdata** after the MPI communication.

With these changes, we could perform the droplet calculation for the active droplets only rather than the whole **nparcelsLocal** slots of **pdata**.

The verification tests show a consistent result compared to those using the latest **gpu-opt** branch (commit: `ff1374`)

sjsprecious committed:
Compress the "holes" in the "pdata" and limit the np loop to the active droplets only in the droplet_driver subroutine

	modified:   comm.F
	modified:   droplet.F
	modified:   init3d.F
	modified:   input.F

sjsprecious committed:
Fix the droplet leak issue

	modified:   comm.F
	modified:   droplet.F

sjsprecious committed:
Merge branch 'gpu-opt' into fix_gpu_bugs

sjsprecious committed:
calculate droplet diagnostic for active droplets only

	modified:   droplet.F

johnmauff commented on src/droplet.F:
```diff
@@ -712,7 +556,45 @@ subroutine droplet_driver(dt,xh,uh,ruh,xf,yh,vh,rvh,yf,zh,mh,rmh,zf,mf,zs,    &
+                                                ! the "pdata" contiguous
+       i = nparcelsActive
+       j = 0
+       do np = 1, nparcelsActive
```
@sjsprecious It appears that you are looping over the entire collection of droplets looking for holes. Why are you not keeping track of the location of all the holes and then filling them in? Right now operation is O(nparcelsActive) versus what it could be which is O(num_holes)

sjsprecious replied:
Thanks for catching this inefficient implementation, @johnmauff ! I have revised the implementation based on your suggestions and the verification results are the same with the latest changes.

johnmauff replied:
Great, thanks for reducing the order of the loop.

johnmauff commented on src/comm.F:
```diff
@@ -12947,8 +12951,8 @@ subroutine comm_droplet_value(holes_ind,num1,num2,pdata,pdata_neighbor)
-      real, intent(inout), dimension(nparcelsLocal,npvals) :: pdata         ! droplet information
-      integer, intent(in), dimension(nparcelsLocal) :: pdata_neighbor       ! array to store the new MPI region 
+      real, intent(inout), dimension(nparcelsLocal,npvals) :: pdata        ! droplet information
+      integer, intent(in), dimension(nparcelsActive) :: pdata_neighbor     ! array to store the new MPI region 
```
Could you clarify if nparcelsActive is really the actual number of active particles or the maximum number of total particles?  If it is the former then why would you dimension arrays a size the should change during execution.

sjsprecious replied:
`nparcelsActive` is the actual number of active droplets per MPI rank and it varies every time step. I think it is fine (and better) to specify the dimension explicitly as long as we update `nparcelsActive` correctly after the call to `droplet_driver`.

sjsprecious replied:
@johnmauff , I have updated the array dimension to a fixed length as we discussed.

johnmauff replied:
Thanks for the code modificaiton.

sjsprecious committed:
Reduce the complexity to search for holes in pdata, following John's suggestion

	modified:   comm.F
	modified:   droplet.F

sjsprecious committed:
Update an array dimension to a fixed length
Simplify the interface in comm.F

	modified:   comm.F
	modified:   droplet.F

johnmauff commented on src/droplet.F:
```diff
@@ -420,101 +387,18 @@ subroutine droplet_driver(dt,xh,uh,ruh,xf,yh,vh,rvh,yf,zh,mh,rmh,zf,mf,zs,    &
 
       ELSEIF (idropmethod.eq.1) THEN  !Implicit backward Euler for the droplet integration
```
@sjsprecious Right now it appears that if the user selects a different droplet time integration method the new MPI implementation will fail right?  I think we should be adding the neighbor array calculation to the rk2_integration time integration method.

sjsprecious replied:
@johnmauff I talked with David once about this issue and he commented that the RK2 solver should be removed sometime in the future. Thus we probably do not want to change it anyway.

johnmauff commented on src/droplet.F:
```diff
@@ -704,18 +549,59 @@ subroutine droplet_driver(dt,xh,uh,ruh,xf,yh,vh,rvh,yf,zh,mh,rmh,zf,mf,zs,    &
 
     call comm_droplet_number(holes_ind,num1,num2,pdata, &
                              pdata_locind,pdata_neighbor)
+    if ( sum(num1) .ne. 0 ) allocate(leave_ind(sum(num1)))
```
@sjsprecious 
The names of the num{1,2} arrays do not convey what they correspond to in the code.  You also use the calculation sum(num1) and sum(num2) quite frequently.  To make the code more readable I would suggest that you rename these arrays:  num1:= Depart, num2:= Arrive, and also define the new variables numDepart = sum(Depart), numArrive=sum(Arrive)

sjsprecious replied:
Thanks for the suggestion. Done.

johnmauff replied:
Thanks this looks good and makes the code much easier to understand

johnmauff commented on src/droplet.F:
```diff
@@ -704,18 +549,59 @@ subroutine droplet_driver(dt,xh,uh,ruh,xf,yh,vh,rvh,yf,zh,mh,rmh,zf,mf,zs,    &
+
+    ! there will be remaining "holes" in the "pdata" only when more
+    ! droplets leave rather than enter the current MPI region
+    if ( sum(num2) .lt. sum(num1) ) then
```
@sjsprecious I am concerned about what you have done with respect to the hole-filling logic.  In particular, it has been placed in two locations.  Once here where the number of arriving droplets is less than the number of departing, and the other is in the comm.F file. Also, I don't think that the update_new_droplet() subroutine really does not belong in the comm.F file. I view this placement as just very confusing for long-term code maintainability.  We probably need a more involved discussion on this topic.

johnmauff replied:
Thanks for moving the hole-filling logic code to a single location in the code

sjsprecious committed:
rename some variables based on John's suggestions

	modified:   comm.F
	modified:   droplet.F

sjsprecious committed:
remove leave_ind for hole filling logic

	modified:   comm.F
	modified:   droplet.F

sjsprecious committed:
Add a new comm module for droplet exchange interface
Move the hole-filling process from droplet_driver to comm_droplet lib

	modified:   Makefile
	modified:   comm.F
	new file:   comm_droplet.F
	modified:   droplet.F

sjsprecious committed:
fix a typo

	modified:   droplet.F

johnmauff reviewed (approved):
@sjsprecious Thanks for all the changes.  This PR now looks good.
